File: pxpoint/cpp/log-services/src/correlation_manager.cpp
#include "correlation_manager.h"
#include <random>
#include <sstream>
#include <iomanip>
#include <chrono>
#include <cstdlib>
#include <fstream>
#include <stdexcept>
#include <yaml-cpp/yaml.h>

namespace logservices {

thread_local std::string CorrelationManager::thread_activity_id_;

CorrelationManager& CorrelationManager::getInstance() {
    static CorrelationManager instance;
    return instance;
}

void CorrelationManager::configure(const CorrelationConfig& config) {
    config_ = config;
}

void CorrelationManager::loadConfigFromYaml(const std::string& yaml_file_path) {
    try {
        YAML::Node config = YAML::LoadFile(yaml_file_path);
        
        if (config["correlation"]) {
            auto corr_config = config["correlation"];
            
            if (corr_config["pipeline_id_prefix"]) {
                config_.pipeline_id_prefix = corr_config["pipeline_id_prefix"].as<std::string>();
            }
            
            if (corr_config["process_id_prefix"]) {
                config_.process_id_prefix = corr_config["process_id_prefix"].as<std::string>();
            }
            
            if (corr_config["activity_id_prefix"]) {
                config_.activity_id_prefix = corr_config["activity_id_prefix"].as<std::string>();
            }
            
            if (corr_config["env_var_pipeline"]) {
                config_.env_var_pipeline = corr_config["env_var_pipeline"].as<std::string>();
            }
            
            if (corr_config["env_var_process"]) {
                config_.env_var_process = corr_config["env_var_process"].as<std::string>();
            }
            
            if (corr_config["auto_generate_pipeline"]) {
                config_.auto_generate_pipeline = corr_config["auto_generate_pipeline"].as<bool>();
            }
            
            if (corr_config["auto_generate_process"]) {
                config_.auto_generate_process = corr_config["auto_generate_process"].as<bool>();
            }
            
            if (corr_config["propagate_to_environment"]) {
                config_.propagate_to_environment = corr_config["propagate_to_environment"].as<bool>();
            }
        }
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to load correlation config from YAML: " + std::string(e.what()));
    }
}

void CorrelationManager::setPipelineId(const std::string& pipeline_id) {
    pipeline_id_ = pipeline_id;
}

std::string CorrelationManager::getPipelineId() const {
    return pipeline_id_;
}

void CorrelationManager::setProcessId(const std::string& process_id) {
    process_id_ = process_id;
}

std::string CorrelationManager::getProcessId() const {
    return process_id_;
}

void CorrelationManager::setActivityId(const std::string& activity_id) {
    thread_activity_id_ = activity_id;
}

std::string CorrelationManager::getActivityId() const {
    return thread_activity_id_;
}

void CorrelationManager::clearActivityId() {
    thread_activity_id_.clear();
}

std::string CorrelationManager::generateDefaultUuid() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    
    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 8; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-4"; // Version 4 UUID
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 12; i++) ss << dis(gen);
    
    return ss.str();
}

std::string CorrelationManager::generateTimestampBasedId(const std::string& prefix) {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;
    
    std::stringstream ss;
    ss << prefix << "-" << time_t << "-" << ms.count() << "-" << generateDefaultUuid().substr(0, 8);
    return ss.str();
}

std::string CorrelationManager::generatePipelineId() {
    if (config_.pipeline_id_generator) {
        return config_.pipeline_id_generator();
    }
    
    return generateTimestampBasedId(config_.pipeline_id_prefix);
}

std::string CorrelationManager::generateProcessId(const std::string& process_type) {
    if (config_.process_id_generator) {
        return config_.process_id_generator(process_type);
    }
    
    if (config_.auto_generate_pipeline && pipeline_id_.empty()) {
        pipeline_id_ = generatePipelineId();
    }
    
    std::stringstream ss;
    if (!pipeline_id_.empty()) {
        ss << pipeline_id_ << "-";
    }
    ss << config_.process_id_prefix << "-" << process_type << "-" << generateDefaultUuid().substr(0, 8);
    return ss.str();
}

std::string CorrelationManager::generateActivityId(const std::string& activity_name) {
    if (config_.activity_id_generator) {
        return config_.activity_id_generator(activity_name);
    }
    
    std::stringstream ss;
    if (!process_id_.empty()) {
        ss << process_id_ << "-";
    }
    ss << config_.activity_id_prefix << "-" << activity_name << "-" << generateDefaultUuid().substr(0, 8);
    return ss.str();
}

void CorrelationManager::loadFromEnvironment() {
    const char* pipeline_env = std::getenv(config_.env_var_pipeline.c_str());
    if (pipeline_env) {
        pipeline_id_ = pipeline_env;
    }
    
    const char* process_env = std::getenv(config_.env_var_process.c_str());
    if (process_env) {
        process_id_ = process_env;
    }
}

void CorrelationManager::saveToEnvironment() const {
    if (!config_.propagate_to_environment) {
        return;
    }
    
    if (!pipeline_id_.empty()) {
        setenv(config_.env_var_pipeline.c_str(), pipeline_id_.c_str(), 1);
    }
    
    if (!process_id_.empty()) {
        setenv(config_.env_var_process.c_str(), process_id_.c_str(), 1);
    }
}

std::string CorrelationManager::getFullCorrelationId() const {
    std::stringstream ss;
    
    if (!pipeline_id_.empty()) {
        ss << "pipeline:" << pipeline_id_;
    }
    
    if (!process_id_.empty()) {
        if (ss.tellp() > 0) ss << "|";
        ss << "process:" << process_id_;
    }
    
    if (!thread_activity_id_.empty()) {
        if (ss.tellp() > 0) ss << "|";
        ss << "activity:" << thread_activity_id_;
    }
    
    return ss.str();
}

std::unordered_map<std::string, std::string> CorrelationManager::getCorrelationContext() const {
    std::unordered_map<std::string, std::string> context;
    
    if (!pipeline_id_.empty()) {
        context["pipeline_id"] = pipeline_id_;
    }
    
    if (!process_id_.empty()) {
        context["process_id"] = process_id_;
    }
    
    if (!thread_activity_id_.empty()) {
        context["activity_id"] = thread_activity_id_;
    }
    
    return context;
}

void CorrelationManager::reset() {
    pipeline_id_.clear();
    process_id_.clear();
    thread_activity_id_.clear();
}

const CorrelationConfig& CorrelationManager::getConfig() const {
    return config_;
}

// ActivityScope implementation
ActivityScope::ActivityScope(const std::string& activity_name) {
    auto& manager = CorrelationManager::getInstance();
    previous_id_ = manager.getActivityId();
    activity_id_ = manager.generateActivityId(activity_name);
    manager.setActivityId(activity_id_);
}

ActivityScope::ActivityScope(const std::string& activity_name, 
                           const std::unordered_map<std::string, std::string>& context)
    : ActivityScope(activity_name) {
    context_ = context;
}

ActivityScope::~ActivityScope() {
    auto& manager = CorrelationManager::getInstance();
    if (previous_id_.empty()) {
        manager.clearActivityId();
    } else {
        manager.setActivityId(previous_id_);
    }
}

std::string ActivityScope::getActivityId() const {
    return activity_id_;
}

void ActivityScope::addContext(const std::string& key, const std::string& value) {
    context_[key] = value;
}

void ActivityScope::addContext(const std::unordered_map<std::string, std::string>& context) {
    for (const auto& [key, value] : context) {
        context_[key] = value;
    }
}

// ProcessScope implementation
ProcessScope::ProcessScope(const std::string& process_type) 
    : created_new_pipeline_(false) {
    auto& manager = CorrelationManager::getInstance();
    
    // Try to load existing correlation from environment
    manager.loadFromEnvironment();
    
    // If no pipeline ID exists and auto-generation is enabled, create a new one
    if (manager.getPipelineId().empty() && manager.getConfig().auto_generate_pipeline) {
        manager.setPipelineId(manager.generatePipelineId());
        created_new_pipeline_ = true;
    }
    
    // Generate process ID for this process type
    if (manager.getConfig().auto_generate_process) {
        process_id_ = manager.generateProcessId(process_type);
        manager.setProcessId(process_id_);
    }
    
    // Save to environment for child processes
    manager.saveToEnvironment();
}

ProcessScope::ProcessScope(const std::string& process_type, 
                         const std::unordered_map<std::string, std::string>& context)
    : ProcessScope(process_type) {
    context_ = context;
}

ProcessScope::~ProcessScope() {
    // Clean up environment variables if we created the pipeline
    if (created_new_pipeline_) {
        auto& manager = CorrelationManager::getInstance();
        unsetenv(manager.getConfig().env_var_pipeline.c_str());
        unsetenv(manager.getConfig().env_var_process.c_str());
    }
}

std::string ProcessScope::getProcessId() const {
    return process_id_;
}

} // namespace logservices

File: pxpoint/log-services-dotnet/src/CorrelationManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading;

namespace LogServices.Correlation
{
    /// <summary>
    /// Configuration for correlation behavior
    /// </summary>
    public class CorrelationConfig
    {
        public string PipelineIdPrefix { get; set; } = "pipeline";
        public string ProcessIdPrefix { get; set; } = "proc";
        public string ActivityIdPrefix { get; set; } = "act";
        public string EnvVarPipeline { get; set; } = "LOG_PIPELINE_ID";
        public string EnvVarProcess { get; set; } = "LOG_PROCESS_ID";
        public bool AutoGeneratePipeline { get; set; } = true;
        public bool AutoGenerateProcess { get; set; } = true;
        public bool PropagateToEnvironment { get; set; } = true;
        
        public Func<string> PipelineIdGenerator { get; set; }
        public Func<string, string> ProcessIdGenerator { get; set; }
        public Func<string, string> ActivityIdGenerator { get; set; }
    }

    /// <summary>
    /// Enhanced correlation manager for multi-process pipeline correlation
    /// Supports hierarchical correlation IDs: Pipeline -> Process -> Activity
    /// Thread-safe with AsyncLocal support for async/await scenarios
    /// </summary>
    public class CorrelationManager
    {
        private static readonly Lazy<CorrelationManager> _instance = 
            new Lazy<CorrelationManager>(() => new CorrelationManager());
        
        public static CorrelationManager Instance => _instance.Value;
        
        // AsyncLocal provides better async/await support than ThreadLocal
        private static readonly AsyncLocal<string> ThreadActivityId = new AsyncLocal<string>();
        
        // Process-level correlation (shared across threads in same process)
        private string _pipelineId = string.Empty;
        private string _processId = string.Empty;
        private CorrelationConfig _config = new CorrelationConfig();
        private readonly object _lock = new object();
        
        private CorrelationManager() { }
        
        // Configuration
        public void Configure(CorrelationConfig config)
        {
            lock (_lock)
            {
                _config = config ?? throw new ArgumentNullException(nameof(config));
            }
        }
        
        public void LoadConfigFromJson(string jsonPath)
        {
            try
            {
                var json = File.ReadAllText(jsonPath);
                var config = JsonSerializer.Deserialize<CorrelationConfig>(json, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                    PropertyNameCaseInsensitive = true
                });
                Configure(config);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Failed to load correlation config from {jsonPath}: {ex.Message}", ex);
            }
        }
        
        public CorrelationConfig GetConfig()
        {
            lock (_lock)
            {
                return _config;
            }
        }
        
        // Pipeline-level correlation (shared across all processes)
        public void SetPipelineId(string pipelineId)
        {
            lock (_lock)
            {
                _pipelineId = pipelineId ?? string.Empty;
            }
        }
        
        public string GetPipelineId()
        {
            lock (_lock)
            {
                return _pipelineId;
            }
        }
        
        // Process-level correlation (unique per process type)
        public void SetProcessId(string processId)
        {
            lock (_lock)
            {
                _processId = processId ?? string.Empty;
            }
        }
        
        public string GetProcessId()
        {
            lock (_lock)
            {
                return _processId;
            }
        }
        
        // Activity-level correlation (within process activities)
        public void SetActivityId(string activityId)
        {
            ThreadActivityId.Value = activityId ?? string.Empty;
        }
        
        public string GetActivityId()
        {
            return ThreadActivityId.Value ?? string.Empty;
        }
        
        public void ClearActivityId()
        {
            ThreadActivityId.Value = string.Empty;
        }
        
        // Generate new correlation IDs
        public string GeneratePipelineId()
        {
            var config = GetConfig();
            if (config.PipelineIdGenerator != null)
            {
                return config.PipelineIdGenerator();
            }
            
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var uuid = Guid.NewGuid().ToString("N")[..8];
            return $"{config.PipelineIdPrefix}-{timestamp}-{uuid}";
        }
        
        public string GenerateProcessId(string processType)
        {
            var config = GetConfig();
            if (config.ProcessIdGenerator != null)
            {
                return config.ProcessIdGenerator(processType);
            }
            
            lock (_lock)
            {
                if (string.IsNullOrEmpty(_pipelineId) && config.AutoGeneratePipeline)
                {
                    _pipelineId = GeneratePipelineId();
                }
                
                var uuid = Guid.NewGuid().ToString("N")[..8];
                return string.IsNullOrEmpty(_pipelineId)
                    ? $"{config.ProcessIdPrefix}-{processType}-{uuid}"
                    : $"{_pipelineId}-{processType}-{uuid}";
            }
        }
        
        public string GenerateActivityId(string activityName)
        {
            var config = GetConfig();
            if (config.ActivityIdGenerator != null)
            {
                return config.ActivityIdGenerator(activityName);
            }
            
            var baseId = GetProcessId();
            var uuid = Guid.NewGuid().ToString("N")[..8];
            
            return string.IsNullOrEmpty(baseId) 
                ? $"{config.ActivityIdPrefix}-{activityName}-{uuid}"
                : $"{baseId}-{activityName}-{uuid}";
        }
        
        // Environment variable integration for cross-process correlation
        public void LoadFromEnvironment()
        {
            var config = GetConfig();
            
            var pipelineEnv = Environment.GetEnvironmentVariable(config.EnvVarPipeline);
            if (!string.IsNullOrEmpty(pipelineEnv))
            {
                SetPipelineId(pipelineEnv);
            }
            
            var processEnv = Environment.GetEnvironmentVariable(config.EnvVarProcess);
            if (!string.IsNullOrEmpty(processEnv))
            {
                SetProcessId(processEnv);
            }
        }
        
        public void SaveToEnvironment()
        {
            var config = GetConfig();
            if (!config.PropagateToEnvironment) return;
            
            var pipelineId = GetPipelineId();
            var processId = GetProcessId();
            
            if (!string.IsNullOrEmpty(pipelineId))
            {
                Environment.SetEnvironmentVariable(config.EnvVarPipeline, pipelineId);
            }
            
            if (!string.IsNullOrEmpty(processId))
            {
                Environment.SetEnvironmentVariable(config.EnvVarProcess, processId);
            }
        }
        
        // Get correlation context for logging
        public string GetFullCorrelationId()
        {
            var parts = new List<string>();
            
            var pipelineId = GetPipelineId();
            if (!string.IsNullOrEmpty(pipelineId))
            {
                parts.Add($"pipeline:{pipelineId}");
            }
            
            var processId = GetProcessId();
            if (!string.IsNullOrEmpty(processId))
            {
                parts.Add($"process:{processId}");
            }
            
            var activityId = GetActivityId();
            if (!string.IsNullOrEmpty(activityId))
            {
                parts.Add($"activity:{activityId}");
            }
            
            return string.Join("|", parts);
        }
        
        public Dictionary<string, object> GetCorrelationContext()
        {
            var context = new Dictionary<string, object>();
            
            var pipelineId = GetPipelineId();
            if (!string.IsNullOrEmpty(pipelineId))
            {
                context["pipeline_id"] = pipelineId;
            }
            
            var processId = GetProcessId();
            if (!string.IsNullOrEmpty(processId))
            {
                context["process_id"] = processId;
            }
            
            var activityId = GetActivityId();
            if (!string.IsNullOrEmpty(activityId))
            {
                context["activity_id"] = activityId;
            }
            
            return context;
        }
        
        // Reset correlation state
        public void Reset()
        {
            lock (_lock)
            {
                _pipelineId = string.Empty;
                _processId = string.Empty;
            }
            ClearActivityId();
        }
    }
    
    /// <summary>
    /// RAII scope for activity-level correlation with context support
    /// </summary>
    public class ActivityCorrelationScope : IDisposable
    {
        private readonly string _previousId;
        private readonly string _activityId;
        private readonly Dictionary<string, object> _context;
        private bool _disposed = false;
        
        public ActivityCorrelationScope(string activityName) : this(activityName, new Dictionary<string, object>())
        {
        }
        
        public ActivityCorrelationScope(string activityName, Dictionary<string, object> context)
        {
            var manager = CorrelationManager.Instance;
            _previousId = manager.GetActivityId();
            _activityId = manager.GenerateActivityId(activityName);
            _context = context ?? new Dictionary<string, object>();
            manager.SetActivityId(_activityId);
        }
        
        public string GetActivityId() => _activityId;
        
        public void AddContext(string key, object value)
        {
            _context[key] = value;
        }
        
        public Dictionary<string, object> GetContext() => new Dictionary<string, object>(_context);
        
        public void Dispose()
        {
            if (!_disposed)
            {
                var manager = CorrelationManager.Instance;
                if (string.IsNullOrEmpty(_previousId))
                {
                    manager.ClearActivityId();
                }
                else
                {
                    manager.SetActivityId(_previousId);
                }
                _disposed = true;
            }
        }
    }
    
    /// <summary>
    /// RAII scope for process initialization with enhanced configuration support
    /// </summary>
    public class ProcessCorrelationScope : IDisposable
    {
        private readonly bool _createdNewPipeline;
        private readonly string _processId;
        private readonly Dictionary<string, object> _context;
        private bool _disposed = false;
        
        public ProcessCorrelationScope(string processType) : this(processType, new Dictionary<string, object>())
        {
        }
        
        public ProcessCorrelationScope(string processType, Dictionary<string, object> context)
        {
            var manager = CorrelationManager.Instance;
            var config = manager.GetConfig();
            _context = context ?? new Dictionary<string, object>();
            
            // Try to load existing correlation from environment
            manager.LoadFromEnvironment();
            
            // If no pipeline ID exists, create a new one
            if (string.IsNullOrEmpty(manager.GetPipelineId()) && config.AutoGeneratePipeline)
            {
                manager.SetPipelineId(manager.GeneratePipelineId());
                _createdNewPipeline = true;
            }
            else
            {
                _createdNewPipeline = false;
            }
            
            // Generate process ID for this process type
            _processId = manager.GenerateProcessId(processType);
            manager.SetProcessId(_processId);
            
            // Save to environment for child processes
            manager.SaveToEnvironment();
        }
        
        public string GetProcessId() => _processId;
        
        public Dictionary<string, object> GetContext() => new Dictionary<string, object>(_context);
        
        public void Dispose()
        {
            if (!_disposed)
            {
                // Clean up if we created the pipeline
                if (_createdNewPipeline)
                {
                    var manager = CorrelationManager.Instance;
                    var config = manager.GetConfig();
                    Environment.SetEnvironmentVariable(config.EnvVarPipeline, null);
                    Environment.SetEnvironmentVariable(config.EnvVarProcess, null);
                }
                _disposed = true;
            }
        }
    }
}